### ISSUE TITLE
Support for GSPO

### ISSUE THREAD

**Issue by YongqiQiao** [OpenRLHF/OpenRLHF#1095]

**Comment by hijkzzz** [comment]
welcome to contribute

**Comment by 122090655** [comment]
Hi, I’m trying to understand how GSPO is implemented in this codebase.

From what I see, it looks like modifying the **PolicyLoss** class in **loss.py** is sufficient to implement GSPO. However, I’m not entirely sure if that alone fully enables GSPO as intended.

Could you please confirm whether modifying PolicyLoss is enough to implement GSPO correctly, or are there other parts of the codebase that also need to be changed?

Thanks!

```python
class PolicyLoss(nn.Module):
    """
    Policy Loss for PPO
    """

    def __init__(self, clip_eps_low: float = 0.2, clip_eps_high: float = 0.2, token_level_loss: bool = True, use_gspo: bool = True) -> None:
        super().__init__()
        self.clip_eps_low = clip_eps_low
        self.clip_eps_high = clip_eps_high
        self.token_level_loss = token_level_loss
        self.use_gspo = use_gspo

    def forward(
        self,
        log_probs: torch.Tensor,
        old_log_probs: torch.Tensor,
        advantages: torch.Tensor,
        action_mask: Optional[torch.Tensor] = None,
    ) -> torch.Tensor:
        if self.use_gspo:
            # GSPO: sequence-level geometric mean ratio
            if action_mask is not None:
                # Only consider valid tokens for geometric mean
                masked_log_probs = log_probs * action_mask
                masked_old_log_probs = old_log_probs * action_mask
                seq_lengths = action_mask.sum(dim=-1, keepdim=True)  # [batch_size, 1]

                # Sum log probs over sequence dimension
                sum_log_probs = masked_log_probs.sum(dim=-1, keepdim=True)  # [batch_size, 1]
                sum_old_log_probs = masked_old_log_probs.sum(dim=-1, keepdim=True)  # [batch_size, 1]

                # Geometric mean: exp(1/|yi| * sum(log_probs))
                mean_log_ratio = (sum_log_probs - sum_old_log_probs) / seq_lengths
                ratio = mean_log_ratio.exp()  # [batch_size, 1]
            else:
                # No masking case
                seq_len = log_probs.size(-1)
                sum_log_probs = log_probs.sum(dim=-1, keepdim=True)
                sum_old_log_probs = old_log_probs.sum(dim=-1, keepdim=True)
                mean_log_ratio = (sum_log_probs - sum_old_log_probs) / seq_len
                ratio = mean_log_ratio.exp()  # [batch_size, 1]

            # GSPO uses sequence-level advantages
            advantages = advantages[:, 0].unsqueeze(1)  # [batch_size, 1]

            # Clipping and loss calculation
            surr1 = ratio * advantages
            surr2 = ratio.clamp(1 - self.clip_eps_low, 1 + self.clip_eps_high) * advantages
            loss = -torch.min(surr1, surr2)  # [batch_size, 1]
            
            # Return batch mean
            loss = loss.mean()
            clip_ratio = torch.lt(surr2, surr1).float().mean()
        else:
            # GRPO: token-level ratio (original implementation)
            ratio = (log_probs - old_log_probs).exp()
            surr1 = ratio * advantages
            surr2 = ratio.clamp(1 - self.clip_eps_low, 1 + self.clip_eps_high) * advantages
            loss = -torch.min(surr1, surr2)
            loss = (
                masked_mean(loss, action_mask, dim=None)
                if self.token_level_loss
                else masked_mean(loss, action_mask, dim=-1).mean()
            )
            clip_ratio = masked_mean(torch.lt(surr2, surr1).float(), action_mask, dim=None)

        return loss, clip_ratio
```

**Comment by YongqiQiao** [comment]
> Hi, I’m trying to understand how GSPO is implemented in this codebase.
> 
> From what I see, it looks like modifying the **PolicyLoss** class in **loss.py** is sufficient to implement GSPO. However, I’m not entirely sure if that alone fully enables GSPO as intended.
> 
> Could you please confirm whether modifying PolicyLoss is enough to implement GSPO correctly, or are there other parts of the codebase that also need to be changed?
> 
> Thanks!
> 
> class PolicyLoss(nn.Module):
>     """
>     Policy Loss for PPO
>     """
> 
>     def __init__(self, clip_eps_low: float = 0.2, clip_eps_high: float = 0.2, token_level_loss: bool = True, use_gspo: bool = True) -> None:
>         super().__init__()
>         self.clip_eps_low = clip_eps_low
>         self.clip_eps_high = clip_eps_high
>         self.token_level_loss = token_level_loss
>         self.use_gspo = use_gspo
> 
>     def forward(
>         self,
>         log_probs: torch.Tensor,
>         old_log_probs: torch.Tensor,
>         advantages: torch.Tensor,
>         action_mask: Optional[torch.Tensor] = None,
>     ) -> torch.Tensor:
>         if self.use_gspo:
>             # GSPO: sequence-level geometric mean ratio
>             if action_mask is not None:
>                 # Only consider valid tokens for geometric mean
>                 masked_log_probs = log_probs * action_mask
>                 masked_old_log_probs = old_log_probs * action_mask
>                 seq_lengths = action_mask.sum(dim=-1, keepdim=True)  # [batch_size, 1]
> 
>                 # Sum log probs over sequence dimension
>                 sum_log_probs = masked_log_probs.sum(dim=-1, keepdim=True)  # [batch_size, 1]
>                 sum_old_log_probs = masked_old_log_probs.sum(dim=-1, keepdim=True)  # [batch_size, 1]
> 
>                 # Geometric mean: exp(1/|yi| * sum(log_probs))
>                 mean_log_ratio = (sum_log_probs - sum_old_log_probs) / seq_lengths
>                 ratio = mean_log_ratio.exp()  # [batch_size, 1]
>             else:
>                 # No masking case
>                 seq_len = log_probs.size(-1)
>                 sum_log_probs = log_probs.sum(dim=-1, keepdim=True)
>                 sum_old_log_probs = old_log_probs.sum(dim=-1, keepdim=True)
>                 mean_log_ratio = (sum_log_probs - sum_old_log_probs) / seq_len
>                 ratio = mean_log_ratio.exp()  # [batch_size, 1]
> 
>             # GSPO uses sequence-level advantages
>             advantages = advantages[:, 0].unsqueeze(1)  # [batch_size, 1]
> 
>             # Clipping and loss calculation
>             surr1 = ratio * advantages
>             surr2 = ratio.clamp(1 - self.clip_eps_low, 1 + self.clip_eps_high) * advantages
>             loss = -torch.min(surr1, surr2)  # [batch_size, 1]
>             
>             # Return batch mean
>             loss = loss.mean()
>             clip_ratio = torch.lt(surr2, surr1).float().mean()
>         else:
>             # GRPO: token-level ratio (original implementation)
>             ratio = (log_probs - old_log_probs).exp()
>             surr1 = ratio * advantages
>             surr2 = ratio.clamp(1 - self.clip_eps_low, 1 + self.clip_eps_high) * advantages
>             loss = -torch.min(surr1, surr2)
>             loss = (
>                 masked_mean(loss, action_mask, dim=None)
>                 if self.token_level_loss
>                 else masked_mean(loss, action_mask, dim=-1).mean()
>             )
>             clip_ratio = masked_mean(torch.lt(surr2, surr1).float(), action_mask, dim=None)
> 
>         return loss, clip_ratio

**Comment by YongqiQiao** [comment]
> Hi, I’m trying to understand how GSPO is implemented in this codebase.
> 
> From what I see, it looks like modifying the **PolicyLoss** class in **loss.py** is sufficient to implement GSPO. However, I’m not entirely sure if that alone fully enables GSPO as intended.
> 
> Could you please confirm whether modifying PolicyLoss is enough to implement GSPO correctly, or are there other parts of the codebase that also need to be changed?
> 
> Thanks!
> 
> class PolicyLoss(nn.Module):
>     """
>     Policy Loss for PPO
>     """
> 
>     def __init__(self, clip_eps_low: float = 0.2, clip_eps_high: float = 0.2, token_level_loss: bool = True, use_gspo: bool = True) -> None:
>         super().__init__()
>         self.clip_eps_low = clip_eps_low
>         self.clip_eps_high = clip_eps_high
>         self.token_level_loss = token_level_loss
>         self.use_gspo = use_gspo
> 
>     def forward(
>         self,
>         log_probs: torch.Tensor,
>         old_log_probs: torch.Tensor,
>         advantages: torch.Tensor,
>         action_mask: Optional[torch.Tensor] = None,
>     ) -> torch.Tensor:
>         if self.use_gspo:
>             # GSPO: sequence-level geometric mean ratio
>             if action_mask is not None:
>                 # Only consider valid tokens for geometric mean
>                 masked_log_probs = log_probs * action_mask
>                 masked_old_log_probs = old_log_probs * action_mask
>                 seq_lengths = action_mask.sum(dim=-1, keepdim=True)  # [batch_size, 1]
> 
>                 # Sum log probs over sequence dimension
>                 sum_log_probs = masked_log_probs.sum(dim=-1, keepdim=True)  # [batch_size, 1]
>                 sum_old_log_probs = masked_old_log_probs.sum(dim=-1, keepdim=True)  # [batch_size, 1]
> 
>                 # Geometric mean: exp(1/|yi| * sum(log_probs))
>                 mean_log_ratio = (sum_log_probs - sum_old_log_probs) / seq_lengths
>                 ratio = mean_log_ratio.exp()  # [batch_size, 1]
>             else:
>                 # No masking case
>                 seq_len = log_probs.size(-1)
>                 sum_log_probs = log_probs.sum(dim=-1, keepdim=True)
>                 sum_old_log_probs = old_log_probs.sum(dim=-1, keepdim=True)
>                 mean_log_ratio = (sum_log_probs - sum_old_log_probs) / seq_len
>                 ratio = mean_log_ratio.exp()  # [batch_size, 1]
> 
>             # GSPO uses sequence-level advantages
>             advantages = advantages[:, 0].unsqueeze(1)  # [batch_size, 1]
> 
>             # Clipping and loss calculation
>             surr1 = ratio * advantages
>             surr2 = ratio.clamp(1 - self.clip_eps_low, 1 + self.clip_eps_high) * advantages
>             loss = -torch.min(surr1, surr2)  # [batch_size, 1]
>             
>             # Return batch mean
>             loss = loss.mean()
>             clip_ratio = torch.lt(surr2, surr1).float().mean()
>         else:
>             # GRPO: token-level ratio (original implementation)
>             ratio = (log_probs - old_log_probs).exp()
>             surr1 = ratio * advantages
>             surr2 = ratio.clamp(1 - self.clip_eps_low, 1 + self.clip_eps_high) * advantages
>             loss = -torch.min(surr1, surr2)
>             loss = (
>                 masked_mean(loss, action_mask, dim=None)
>                 if self.token_level_loss
>                 else masked_mean(loss, action_mask, dim=-1).mean()
>             )
>             clip_ratio = masked_mean(torch.lt(surr2, surr1).float(), action_mask, dim=None)
> 
>         return loss, clip_ratio

Thanks，i found another approach，https://github.com/huggingface/trl/commit/03034317d0be0c259c315f5ffad71be138c17d2c
this may be more concise
